### DISCUSSION TITLE
DEC2HEX...but not the usual way

### DISCUSSION THREAD

**Discussion by DRBAUS** [RomanKubiak/ctrlr#129] [General]
Okay...I might want to integrate a little bit of sample editing in the panel that I am working on and I've already been reading up on it and I think that the dec2hex won't work for this piece of gear. 

Option 1 is to include a uiCombo with most common values that I use for editing but I wonder if it is possible to get a formula for the way that this Ensoniq unit handles the sysex. I will explain.

The value range is 0 - FFFFFFFF

The sysex command for sample start = 0 is:
{0xF0, 0x0F, 0x03, 0x00, 0x11, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x20, 0x15, 0x00, 0x00, 0x00, 0x00, 0xF7}

The way this unit counts is in steps of 64 (3F) so value = 63 is:
{0xF0, 0x0F, 0x03, 0x00, 0x11, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x20, 0x15, 0x00, 0x00, 0x00, 0x3F, 0xF7}
value = 64 is:
{0xF0, 0x0F, 0x03, 0x00, 0x11, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x20, 0x15, 0x00, 0x00, 0x01, 0x00, 0xF7}
value = 127 is:
{0xF0, 0x0F, 0x03, 0x00, 0x11, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x20, 0x15, 0x00, 0x00, 0x01, 0x3F, 0xF7}
value = 128 is:
{0xF0, 0x0F, 0x03, 0x00, 0x11, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x20, 0x15, 0x00, 0x00, 0x02, 0x00, 0xF7}
 
etc. until 02 reaches 3F
value = 4095 is
{0xF0, 0x0F, 0x03, 0x00, 0x11, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x20, 0x15, 0x00, 0x00, 0x3F, 0x3F, 0xF7}
value = 4096 is
{0xF0, 0x0F, 0x03, 0x00, 0x11, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x20, 0x15, 0x00, 0x01, 0x00, 0x00, 0xF7}

and going on like this.

How would I approach this in a method? Is it possible to convert the (decimal) user input of a uiLabel to a sysex table?
I have no idea how to look at this so any input is welcome. 

There's no rush but if this is possible it means that the panel could also become a sample editor and that would be cool.

Thanks in advance,

BAUS

**Comment by Ted-juh** [comment]
0x01 0x00 0x00 0x00 = 262144?
0x3F 0x3F 0x3F 0x3F = 16777215?
Then 16777216 = 0x3F 0x3F 0x3F 0x40?

Could you write down the Bytes in HEX for FFFFFFFF?

**Reply by DRBAUS** [reply]
Good question! I don't know.

I went with what was mentioned in the sysex manual here....appears the largest value this particular sampler can have is 1044088 (0x03 0x3E 0x39 0x38)

**Comment by dobo365** [comment]
Hi,
  create a uiImageSlider or to test a simple uiSlider and add a method like:

```
--
-- Called when a modulator value changes
-- @mod   http://ctrlr.org/api/class_ctrlr_modulator.html
-- @value    new numeric value of the modulator
--
test = function(--[[ CtrlrModulator --]] mod, --[[ number --]] value, --[[ number --]] source)

	local iByte1 = math.modf(value/4096)
	local iByte2 = math.modf((value - iByte1 * 4096)/128)
	local iByte3 = value - iByte1*4096 - iByte2 * 128

	msg = string.format("F0 0F 03 00 11 00 00 00 00 00 01 20 15 00 %.2x %.2x %.2x F7", iByte1, iByte2, iByte3)	
	console(msg)

end
```

Doesn't give the right values yet so need to adapt the calculations of the different bytes but you get the philosophy.
Never forget that hexa is just numbers as decimal ;-)
So, separate the calculation from the string message.

**Comment by dobo365** [comment]
I took it back this evening and by re-reading your post I saw you said it counts by 64.
The good code is thus same as above but with the 128 replaced by 64. Works perfect ;-)
```
test = function(--[[ CtrlrModulator --]] mod, --[[ number --]] value, --[[ number --]] source)

	local iByte1 = math.modf(value/4096)
	local iByte2 = math.modf((value - iByte1 * 4096)/64)
	local iByte3 = value - iByte1*4096 - iByte2 * 64

	msg = string.format("F0 0F 03 00 11 00 00 00 00 00 01 20 15 00 %.2x %.2x %.2x F7", iByte1, iByte2, iByte3)	
	console(msg)
--	panel:sendMidiMessageNow(CtrlrMidiMessage(msg))

end
```
test(panel:getModulatorByName("modulator-1"), 63)
F0 0F 03 00 11 00 00 00 00 00 01 20 15 00 00 00 3f F7

test(panel:getModulatorByName("modulator-1"), 64)
F0 0F 03 00 11 00 00 00 00 00 01 20 15 00 00 01 00 F7

test(panel:getModulatorByName("modulator-1"), 127)
F0 0F 03 00 11 00 00 00 00 00 01 20 15 00 00 01 3f F7

test(panel:getModulatorByName("modulator-1"), 128)
F0 0F 03 00 11 00 00 00 00 00 01 20 15 00 00 02 00 F7

test(panel:getModulatorByName("modulator-1"), 4095)
F0 0F 03 00 11 00 00 00 00 00 01 20 15 00 00 3f 3f F7

test(panel:getModulatorByName("modulator-1"), 4096)
F0 0F 03 00 11 00 00 00 00 00 01 20 15 00 01 00 00 F7

/goodweather

**Comment by DRBAUS** [comment]
.....and I thought you did that on purpose for me to find out it needed to be 64....lol 

I tested it, saw the result and thought....hey, that probably needs to be 64 instead of 128.

Now with the uiImageSlider or uiSlider we can only have a maximum value of 65536....at least that is in the version of Ctrlr that I am working with 5.3.196 Is this any different on the most recent update?

If not then I will ask about part 2 of the original question: How do we get the user input of the uiTextLabel as the value to do the above calculations on? The largest decimal value that I read in this sampler is 1044088 (H03 H3E H39 H38).

Thanks a lot. This is really becoming more fun everyday. :D

BAUS

**Reply by DRBAUS** [reply]
Oh....that was easier than I thought.

`	panel:getModulatorByName("TestLabel"):getProperty(value)
	
	local iByte1 = math.modf(value/4096)
	local iByte2 = math.modf((value - iByte1 * 4096)/64)
	local iByte3 = value - iByte1*4096 - iByte2 * 64

	msg = string.format("F0 0F 03 00 11 00 00 00 00 00 01 20 15 00 %.2x %.2x %.2x F7", iByte1, iByte2, iByte3)	
	console(msg)
	panel:sendMidiMessageNow(CtrlrMidiMessage(msg))

end`

**Comment by Ted-juh** [comment]
How's this BAUS?
``

test = function(mod, value, source)

	value = L(panel:getComponent("TestLabel"):getText())


	local iByte1 = math.modf(value/262144)
	local iByte2 = math.modf((value - iByte1*262144)/4096)
	local iByte3 = math.modf((value - (iByte1*262144 + iByte2*4096))/64)
	local iByte4 = math.modf(value - (iByte1*262144 + iByte2*4096 + iByte3*64))


	msg = string.format("F0 0F 03 00 11 00 00 00 00 00 01 20 15 %.2x %.2x %.2x %.2x F7", iByte1, iByte2, iByte3, iByte4)	
	console(msg)
--	panel:sendMidiMessageNow(CtrlrMidiMessage(msg))

end
``

Geen dank he? :)

**Comment by dobo365** [comment]
Well...I have a full day time job and could only do it quickly in the morning as I knew the answer but didn't have the time to test it further.

You should move to 5.3.201 which is the latest stable version.
@Tedjuh1976 gave you the code to handle higher values

Now, you should really check and understand the synth (this is the key when designing panels and it is possible even without having the synth). I mean:
- what kind of controller will go to such high values in steps of 1? It is very difficult to adjust a Slider when you go to high values in steps of 1
- if you have bigger steps between values then you can just divide/multiply by that amount and reduce the values of your sliders (either in the code or with the properties "_Expression to evaluate when.._."). E.g. with  100 values but a step of 1270 I can go to 127000...
- do you need a slider to enter that high value? Shouldn't you instead use uiCombo or ...?

Raising all those questions is where your panel design skills should shine in ;-)

**Comment by dnaldoog** [comment]
Here's an alternative to goodweather's and Tedjuh's excellent suggestions using [JUCE BigInteger class ](https://docs.juce.com/master/classBigInteger.html#ac78d23b9c93713430758df01c0a68f48):

This code will cover a full 16 bit unsigned value 0 ~ 65,535 or 0xFFFF as in the manual _page 5_ **16 BIT Word Format**:

<pre>
ensoniq16BitMask = function(--[[ CtrlrModulator --]] mod, --[[ number --]] value, --[[ number --]] source)
local bv=BigInteger(value)
	local iByte1 = bv:getBitRangeAsInt(12,4)
	local iByte2 = bv:getBitRangeAsInt(6,6)
	local iByte3 = bv:getBitRangeAsInt(0,6)
	local msg = string.format("F0 0F 03 00 11 00 00 00 00 00 01 20 15 00 %.2x %.2x %.2x F7", iByte1, iByte2, iByte3)	
--	panel:sendMidiMessageNow(CtrlrMidiMessage(msg))
end
</pre>

This code will cover a 12 bit unsigned value 0 ~ 4,095 or 0xFFF as in the manual _page 5_ **12 BIT Word Format**:

<pre>
ensoniq12BitMask = function(--[[ CtrlrModulator --]] mod, --[[ number --]] value, --[[ number --]] source)
local bv=BigInteger(value)
	local iByte2 = bv:getBitRangeAsInt(6,6)
	local iByte3 = bv:getBitRangeAsInt(0,6)
	local msg = string.format("F0 0F 03 00 11 00 00 00 00 00 01 20 15 00 00 %.2x %.2x F7",  iByte2, iByte3)	
--	panel:sendMidiMessageNow(CtrlrMidiMessage(msg))
end
</pre>

----

[Ensoniq_EPS16Plus_External_Command_Specification_130.zip](https://github.com/RomanKubiak/ctrlr/files/8339567/Ensoniq_EPS16Plus_External_Command_Specification_130.zip)
